contracts/StorageManager.sol

```diff
@@ -172,177 +158,187 @@ contract StorageManager {
-            numberOfPeriodsDeposited.mul(billingPeriod) <= offer.maximumDuration &&
-            numberOfPeriodsDeposited <= MAX_UINT64,
-            "StorageManager: total period exceeds maximumDuration"
-        );
-        //NO_OVERFLOW reasoning: verified above
-        now.add(numberOfPeriodsDeposited * billingPeriod); // overFlow check. If this doesn't pass, the duration of the offer overflows MAX_UINT64 and the contract may deadlock. REF_DURATION
-        numberOfPeriodsDeposited.mul(billingPrice); // overFlow check. If this doesn't pass, we might have issues transfering the maximum amount REF_MAX_TRANSFER
+        require(agreement.lastPayoutDate == 0, "StorageManager: Agreement already active");
+
+        uint64 billingPrice = offer.billingPlans[billingPeriod];
+        require(billingPrice != 0, "StorageManager: billing price doesn't exist for Offer");
+
+        // Adding to previous availableFunds as the agreement could have been expired
+        // and Consumer is reactivating it, so in order not to loose any previous funds.
+        agreement.availableFunds = agreement.availableFunds.add(msg.value);
+        require(agreement.availableFunds >= size * billingPrice, "StorageManager: funds deposited has to be for at least one billing period");
+
+        agreement.size = size;
         agreement.billingPrice = billingPrice;
         agreement.billingPeriod = billingPeriod;
-        agreement.numberOfPeriodsDeposited = uint64(numberOfPeriodsDeposited);
-        agreement.startDate = uint128(now);
+        agreement.lastPayoutDate = uint128(now);
+
+        offer.availableCapacity = uint128(offer.availableCapacity.sub(size));
+
         emit NewAgreement(
+            agreementReference,
             dataReference,
             msg.sender,
             provider,
             size,
             billingPeriod,
-            msg.value
+            billingPrice,
+            agreement.availableFunds
         );
     }
 
     /**
-    @notice extend the duration of the agreement.
-    @dev any safeMath operations in this function don't cause a deadlock, as a possible revert just means we can't prolong the agreement for the desired duration\
-    @param dataReference the reference to the to-be-pinned file
+    >> FOR CONSUMER
+    @notice deposits new funds to the Agreement.
+    @dev
+        - depositing funds to Agreement that is linked to terminated Offer is not possible
+    @param dataReference data reference where should be deposited funds.
     @param provider the address of the provider of the Offer.
     */
     function depositFunds(bytes32[] memory dataReference, address provider) public payable {
-        bytes32 agreementReference = getAgreementReference(dataReference);
+        bytes32 agreementReference = getAgreementReference(dataReference, msg.sender);
         Offer storage offer = offerRegistry[provider];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
```

Comment:
This line is used in many functions. Could probably be extracted to a "modifier" such as onlyOwner and then add the modifier to the functions. we can tackle this in future PR for optimizations  if preferable.

Reply:
Good suggestion! Done